### PR TITLE
Consensus refactor

### PR DESCRIPTION
**Pull request by bachase**
This is the next phase of consensus refactoring in support of a simulation and testing framework.  The last refactor #2040 focused on minimal changes to get decent unit test coverage of the generic consensus algorithm.  This review's changesets focus on slimming down and simplifying the generic consensus code, including:

* Move the option of `validation` from generic `Consensus` to `RCLConsensus`, since generic consensus doesn't directly deal with validating the consensus ledger.
* Move exception handling to `RCLConsensus`,  since it was added sometime in the past to catch `SHAMap` missing node edge cases.
* Refactor `gotTxSet` and `gotTxSetInternal`, since receiving a peer's transaction set requires a different response than handling our own transaction set.  
* Introduce a `Consensus::Mode` enum to better represent the interaction between  `wrongLCL` and `proposing_` .
* Rename `Consensus::State` to `Consensus::Phase` and eliminate the `processing` phase.  Instead, a consensus round terminates in the `accept` phase. The choice to dispatch that result to another thread is now completely opaque to the generic consensus algorithm.  Instead, until a call to `startRound`, all other calls into consensus do nothing while in the `accept` phase.
* Introduce a `Consensus::Result` object to hold the outcomes of consensus.  This replaces two optionals for the node's transaction set and position with a single optional to a `Result` holding both, along with the disputes.   This structurally enforces the invariant that both are created together and that no disputes are created until a node takes an initial position.

I also took some time to write a [high-level overview](https://bachase.github.io/ripd1340/rippled/consensus_and_validation.html) of the consensus algorithm.  I suggest reviewer's consult that before reviewing the code changes to help understand the final design.  I welcome feedback on the documentation as well and hope I didn't butcher quickbook too much ;).  

Top-level documentation preview https://bachase.github.io/ripd1340/.

### CODE REVIEW

**Comment by bachase** [comment]
@ximinez I'd appreciate your feedback on the quickbook documentation I added in 6317d9e350b1cad1f5f04391588f139e2e6c1658.  I struggled with the proper sectioning and went with the least worst result.

File: docs/consensus.qbk [hunk 1]
```diff
@@ -0,0 +1,651 @@
...
+state.  For the Ripple network, the shared state is the current ledger--account
+information, account balances, order books and other financial data.  We will
+refer to shared distributed state as a /ledger/ throughout the remainder of this
+document.
+
+[$images/consensus/ledger_chain.png [width 50%] [height 50%] ]
+
+As shown above, new ledgers are made by applying a set of transactions to the
+prior ledger.  For the Ripple network, transactions include payments,
+modification of account settings, updates to offers and more. For Git,
```
Comment (bachase) [inline docs/consensus.qbk]: The `For git ...` is from an earlier markdown iteration.  I will remove.

File: docs/consensus.qbk [hunk 2]
```diff
@@ -0,0 +1,651 @@
...
+Calls to =peerProposal= or =gotTxSet= that reflect new positions will generate
+disputed transactions for any new disagreements and will update the peer's vote
+for all disputed transactions.
+
+A call to =timerEntry= first checks that the node is working from the correct
+prior ledger. If not, the node  will update the mode and request the correct
+ledger.  Otherwise, the node updates the node's position and considers whether
+to switch to the =Accepted= phase and declare consensus reached.  However, at
+least =LEDGER_MIN_CONSENSUS= time must have elapsed before doing either.  This
+allows peer's an opportunity to take an initial position and share it.
```
Comment (ximinez) [inline docs/consensus.qbk]: `allows peers an opportunity` (No possessive.)

File: docs/consensus.qbk [hunk 3]
```diff
@@ -0,0 +1,651 @@
...
+Conversely, if the diagram reflected a later call to =timerEntry= that occurs in
+the stuck region with a threshold of say 95%, our node would remove transactions
+2 and 3 from its candidate set and send an updated position.  Likewise, all the
+other peers would end up with only transaction 1 in their position.
+
+Lastly, if our node were not in the proposing mode, it would not include its own
+vote and just take the majority (>50%) position of its peers. In this example,
+our node would maintain its position of transactions 1, 2 and 3.
+
+['Checking Consensus]
```
Comment (ximinez) [inline docs/consensus.qbk]: I like these subsections for the actions / transitions, and think they could do well under the other states.

File: docs/consensus.qbk [hunk 4]
```diff
@@ -0,0 +1,651 @@
...
+   using ID = ...;
+   ID const & id() const;
+
+   //... implementation specific
+};
+```
+
+[heading Transaction Set]
+The transaction set type =TxSet= represents a set of =Tx=s that are collectively
+under consideration by consensus. A =TxSet= can be compared against other =TxSet=s
```
Comment (ximinez) [inline docs/consensus.qbk]: As above, the `=Tx=s` and `=TxSet=s` don't get interpreted correctly here. I don't have a suggestion on how to rephrase.

File: docs/consensus.qbk [hunk 5]
```diff
@@ -0,0 +1,651 @@
...
+
+  Json::Value getJson() const;
+
+  //... implementation specific
+};
+```
+[heading Generic Consensus Interface]
+
+Following the
+[@https://en.wikipedia.org/wiki/Curiously/recurring/template_pattern CRTP]
```
Comment (ximinez) [inline docs/consensus.qbk]: Broken link. Try `https://en.wikipedia.org/wiki/Curiously_recurring_template_pattern`

File: docs/consensus.qbk [hunk 6]
```diff
@@ -0,0 +1,651 @@
...
+algorithm.  Below are excerpts of the generic consensus implementation and of
+helper types that will interact with the concrete implementing class.
+
+```
+
+// Represents our proposed position or a peer's proposed position template
+<class NodeID_t, class LedgerID_t, class Position_t> class ConsensusProposal;
+
+// Represents a transction under dispute this round template <class Tx_t, class
+NodeID_t> class DisputedTx;
```
Comment (ximinez) [inline docs/consensus.qbk]:
```
// Represents a transaction under dispute this round
template <class Tx_t, class NodeID_t> class DisputedTx;
```

**Comment by JoelKatz** [comment]
:neckbeard::+1: Code passes functional testing. Documentation looks correct. Code looks good. I suggested some tiny changes to the documentation and some whitespace fixes. They are not particularly important.

**Comment by ximinez** [comment]
Nit: There are several lines in consensus.qbk with trailing whitespace. I removed them for you. https://github.com/ximinez/rippled/commit/980249a86f63ab54f4011653baeefa3114ae9732

**Comment by bachase** [comment]
Updated documentation https://bachase.github.io/ripd1340/rippled/consensus_and_validation.html and fixed missing newlines.

File: src/test/csf/BasicNetwork.h [hunk 7]
```diff
@@ -473,24 +470,20 @@ class BasicNetwork
...
+    step_for(std::chrono::duration<Period, Rep> const& amount);
 };
 
 //------------------------------------------------------------------------------
 
 template <class Peer>
-BasicNetwork<Peer>::queue_type::queue_type(
-        qalloc const& alloc)
-    : alloc_ (alloc)
+BasicNetwork<Peer>::queue_type::queue_type(qalloc const& alloc) : alloc_(alloc)
```
Comment (ximinez) [inline src/test/csf/BasicNetwork.h]: Nit: this line is longer than we usually allow. I'd put the initializer back on its own line.

File: src/ripple/app/consensus/RCLConsensus.cpp [hunk 8]
```diff
@@ -249,232 +214,242 @@ RCLConsensus::acquireTxSet(RCLTxSet::ID const & setId)
...
-            now,
-            nodeID_ });
+    return Result(
+        std::move(initialSet),
+        RCLCxPeerPos::Proposal{initialLedger->info().parentHash,
+                               RCLCxPeerPos::Proposal::seqJoin,
+                               setHash,
+                               closeTime,
+                               app_.timeKeeper().closeTime(),
+                               nodeID_});
```
Comment (bachase) [inline src/ripple/app/consensus/RCLConsensus.cpp]: I was using @seelabs clang-format and this is what popped out.  I'll try to change it to be less indent heavy.

Comment (seelabs) [inline src/ripple/app/consensus/RCLConsensus.cpp]: @bachase For some reason I can't view the code (??). Can you save the snippet so we can try to tweak the format params to do what we want?

Comment (bachase) [inline src/ripple/app/consensus/RCLConsensus.cpp]: @seelabs This is the spot https://github.com/bachase/rippled/blob/consensus-refactor/src/ripple/app/consensus/RCLConsensus.cpp#L340-L347

Comment (seelabs) [inline src/ripple/app/consensus/RCLConsensus.cpp]: @bachase Thanks. It looks like the brace initializers isn't handled like we want. If I replace that with parens I get: ```
return Result(
        std::move(initialSet),
        RCLCxPeerPos::Proposal(
            initialLedger->info().parentHash,
            RCLCxPeerPos::Proposal::seqJoin,
            setHash,
            closeTime,
            app_.timeKeeper().closeTime(),
            nodeID_));
```

Comment (bachase) [inline src/ripple/app/consensus/RCLConsensus.cpp]: Got it.  I'll go with that for now.  Thanks!

Comment (seelabs) [inline src/ripple/app/consensus/RCLConsensus.cpp]: I couldn't figure out how to fix the `.clang_format` to handle this correctly. I'll try again later, but it wasn't obvious. 

Side note: I don't want the formatting tool driving what language features we use. If the formatting tool is messing up, I wouldn't use it for that snippet for now.

**Comment by bachase** [comment]
Rebased on 0.70.0-b3

**Comment by bachase** [comment]
Squashed

**Comment by bachase** [comment]
Rebased on 0.70.0-b4

**Comment by scottschurr** [comment]
Merged to rippled develop as 00c60d408a887d8a986db81afbb5ead121e8310c.

**Pull request closed without merging**